### PR TITLE
[Expanded Weights] fix loss reduction

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80944
* #80943
* #80895
* __->__ #80892

Two changes in here:
(1) Changes `call_for_per_sample_grads` to be curried. Old call looks like:
`call_for_per_sample_grads(module, batch_size, args, kwargs)`
New call looks like:
`call_for_per_sample_grads(module, batch_size, loss_reduction=loss_reduction)(args, kwargs)`

(2) Adds the ability to specify a loss reduction, to match what is done in Opacus. Opacus has a more complete explanation but essentially, they want the per sample gradient behavior to match what is happens in a for loop with a single example. This gets messed up if you use a mean reduction at the end since in a batch that ends up scaling all the grad_outputs by 1/batch_size, so we offset that by scaling all the grad_samples by batch_size if the loss_reduction is mean

### CODE REVIEW

File: torch/nn/utils/_per_sample_grad.py [hunk 1]
```diff
@@ -1,29 +1,30 @@
+import functools
+
 import torch
 from torch.nn.utils._stateless import functional_call
 from torch.nn.utils._expanded_weights.expanded_weights_impl import ExpandedWeight
 
 # dependency on `functional_call` means that this can't be exposed in utils
 # without creating circular dependency
-def call_for_per_sample_grads(module, batch_size, args, kwargs=None):
+def call_for_per_sample_grads(module, batch_size, loss_reduction="mean"):
```
Comment (zou3519) [inline torch/nn/utils/_per_sample_grad.py:9]: I don't really like having a loss_reduction argument because the loss is not guaranteed to have a loss_reduction argument (in general) but also because nll_loss's mean argument is different from other losses.

[(docs)](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss) nll_loss's mean argument doesn't mean "divide by batch_size", it means divide by the sum of the relevant weights. So if someone uses nll_loss they won't actually be able to reproduce the semantics of a for-loop.


To check my understanding about the motivation, a forward+backward pass looks like the following:
```
model = ...
targets = ...
criterion = CrossEntropyLoss(reduction='mean')
result = call_for_per_sample_grads(model, batch_size, input)
loss = criterion(result, targets)
loss.backward()
```
what happens if you move the criterion call to inside the model? Does that give the desired output?

Comment (zou3519) [reply torch/nn/utils/_per_sample_grad.py:9]: I see, thanks for the detailed explanation. This seems like a fundamental limitation of the current design; there is no real way to detect that a user is doing "batch-dependent operations" (in this case we are talking about dividing the input by 1/batch_size, but another example is batch_norm) so we resolve this via user intervention by asking them to provide us a string

File: torch/nn/utils/_per_sample_grad.py [hunk 2]
```diff
@@ -1,29 +1,31 @@
...
-    call_for_per_sample_grads(module, batch_size, args, kwargs=None) -> Tensor
-    Invoked just like a forward pass, ``call_for_per_sample_grads`` will produce the same
-    forward result. Then, when backward is invoked, the parameters of ``module``
-    will have a ``grad_sample`` field populated with the per sample gradients
-    instead of the regular gradients
+    call_for_per_sample_grads(module, batch_size, loss_reduction="sum")
+    ``call_for_per_sample_grads`` returns a function that is invoked like the forward
+    function of ``module`` and will produce the same result. Then, when backward is invoked,
+    the parameters of ``module`` will have a ``grad_sample`` field populated with the per sample
+    gradients instead of the regular gradients
```
Comment (zou3519) [inline torch/nn/utils/_per_sample_grad.py:15]: What is the motivation for changing `call_for_per_sample_grads` to be curried?

File: torch/nn/utils/_expanded_weights/expanded_weights_utils.py [hunk 3]
```diff
@@ -52,18 +54,34 @@ def _check_and_unexpand_args(func, expanded_args, expanded_kwargs):
...
+        return res
+
     unpacked = unpack_expanded_weight_or_tensor(maybe_expanded_weight)
     if isinstance(maybe_expanded_weight, ExpandedWeight):
         if hasattr(unpacked, "grad_sample") and unpacked.grad_sample is not None:
-            unpacked.grad_sample = unpacked.grad_sample + per_sample_grad_fn(unpacked)
+            unpacked.grad_sample = unpacked.grad_sample + loss_reduction_wrapper(unpacked, grad_output)
         else:
-            unpacked.grad_sample = per_sample_grad_fn(unpacked)
+            unpacked.grad_sample = loss_reduction_wrapper(unpacked, grad_output)
```
Comment (zou3519) [inline torch/nn/utils/_expanded_weights/expanded_weights_utils.py:84]: I'm a bit confused -- why do we need to accept grad_output? Could we just do the scaling like the following:
```
def set_grad_sample_if_exists(maybe_expanded_weight, per_sample_grad_fn):
    unpacked = unpack_expanded_weight_or_tensor(maybe_expanded_weight)
    if isinstance(maybe_expanded_weight, ExpandedWeight):
        grad_sample_contribution = maybe_scale_by_batch_size(per_sample_grad_fn(unpacked))
        if hasattr(unpacked, "grad_sample") and unpacked.grad_sample is not None:
            unpacked.grad_sample = unpacked.grad_sample + grad_sample_contribution
        else:
            unpacked.grad_sample = grad_sample_contribution
```

File: test/test_expanded_weights.py [hunk 4]
```diff
@@ -131,38 +139,59 @@ def test_sum_over_all_but_batch_and_last_n(self, device):
...
+        # get per sample grads with for loop
+        func = partial(run_op, op)
+
+        per_sample_grad = for_loop_per_sample_grad(batch_size, reduction, input, func, *args, **kwargs)
+
+        # check equality
+        self.assertEqual(len(per_sample_grad), len(expanded_weight_grad))
+        if loss_reduction == "mean":
+            expanded_weight_grad = expanded_weight_grad[1:]
+            per_sample_grad = per_sample_grad[1:]
```
Comment (zou3519) [inline test/test_expanded_weights.py:169]: Why the [1:]?

File: test/test_expanded_weights.py [hunk 5]
```diff
@@ -131,38 +139,59 @@ def test_sum_over_all_but_batch_and_last_n(self, device):
...
+        for (result_grad, expected_grad) in zip(expanded_weight_grad, per_sample_grad):
+            if result_grad is None:
+                result_grad = torch.zeros_like(expected_grad)
+            self.assertEqual(result_grad, expected_grad)
```
Comment (zou3519) [inline test/test_expanded_weights.py:173]: Is the implication that sometimes ExpandedWeights will return an all-zeros gradient when regular autograd would have returned a None gradient?

**Review by zou3519 — COMMENTED** [review]
Code looks correct to me, I have some minor questions about the implementation

File: torch/nn/utils/_per_sample_grad.py (hunk 1 above)
Comment (zou3519) [reply torch/nn/utils/_per_sample_grad.py]: Resolved offline: we're just going to default loss reduction to 'sum'. There's a "long term expanded weights" design to fix this, but not much we can do in the short term.

File: test/test_expanded_weights.py (hunk 4 above)
Comment (samdow) [reply test/test_expanded_weights.py:169]: The 0th element corresponds to the input. Because we aren't actually updating `grad_output`, the derivative wrt the input is going to be different between the for loop version and the expanded weights version because the input is a vanilla tensor and so won't be scaled

File: test/test_expanded_weights.py (hunk 5 above)
Comment (samdow) [reply test/test_expanded_weights.py:173]: The thought was that this might come up if the weight was unused and torch.autograd.grad produced a tensor of 0s but ExpandedWeights produced None (since we check if it's None to see if we should make the first tensor or accumulate). In practice though because these test one function at a time, this actually doesn't come up so I was able to remove these lines

File: test/test_expanded_weights.py (hunk 4 above)
Comment (zou3519) [reply test/test_expanded_weights.py:169]: Ah I forgot about that. Might be good to add a comment about it

**Review by zou3519 — APPROVED** [review]

File: torch/nn/utils/_per_sample_grad.py (hunk 2 above)
Comment (zou3519) [reply torch/nn/utils/_per_sample_grad.py:15]: Good point, thanks for the explanation
